### PR TITLE
feat: add CHGNet PyG backend with pre-trained MatPES-PBE and r2SCAN potentials

### PR DESCRIPTION
**Pull request by bowen-bd**
# CHGNet-PyG: DGL ↔ PyG Prediction Parity Report

## Overview

This PR adds a PyTorch Geometric (PyG) backend implementation of CHGNet and releases
two pre-trained potentials whose weights were **directly transferred** from the existing
DGL checkpoints — no retraining was performed.

| Model | Functional | DGL checkpoint | PyG checkpoint |
|---|---|---|---|
| CHGNet-PES-MatPES-PBE-2025.2.10 | PBE | [`materialyze/CHGNet-PES-MatPES-PBE-2025.2.10`](https://huggingface.co/materialyze/CHGNet-PES-MatPES-PBE-2025.2.10) | [`BowenD-UCB/CHGNet-PyG-MatPES-PBE-2025.2.10`](https://huggingface.co/BowenD-UCB/CHGNet-PyG-MatPES-PBE-2025.2.10) |
| CHGNet-PES-MatPES-r2SCAN-2025.2.10 | r2SCAN | [`materialyze/CHGNet-PES-MatPES-r2SCAN-2025.2.10`](https://huggingface.co/materialyze/CHGNet-PES-MatPES-r2SCAN-2025.2.10) | [`BowenD-UCB/CHGNet-PyG-MatPES-r2SCAN-2025.2.10`](https://huggingface.co/BowenD-UCB/CHGNet-PyG-MatPES-r2SCAN-2025.2.10) |

The PyG checkpoints are currently hosted under the `BowenD-UCB` HuggingFace account and
can be transferred to the `materialyze` org if @shyuep agrees on this.  Until then, load
them with the full owner-qualified path:

```python
import matgl

# r2SCAN (recommended)
pot = matgl.load_model("BowenD-UCB/CHGNet-PyG-MatPES-r2SCAN-2025.2.10")

# PBE
pot = matgl.load_model("BowenD-UCB/CHGNet-PyG-MatPES-PBE-2025.2.10")
```

Once transferred to `materialyze`, bare-name loading will work:

```python
pot = matgl.load_model("CHGNet-PyG-MatPES-r2SCAN-2025.2.10")
```

A usage example covering static prediction, ionic relaxation, molecular dynamics, and
training/fine-tuning is provided in
[`examples/CHGNet-PyG Foundation Potential.ipynb`](examples/CHGNet-PyG%20Foundation%20Potential.ipynb).

---

## Architecture Correspondence

The PyG implementation (`src/matgl/models/_chgnet_pyg.py`) is a faithful port of the
DGL CHGNet (`src/matgl/models/_chgnet.py`).  Key structural elements map one-to-one:

| Component | DGL | PyG |
|---|---|---|
| Atom graph | `dgl.DGLGraph` node/edge data | `torch_geometric.data.Data` |
| Line graph (three-body) | `_create_directed_line_graph` (DGL op) | `create_directed_line_graph_pyg` (vectorised) |
| Graph convolution | `CHGNetConv` (`_graph_convolution_dgl.py`) | `CHGNetConv` (`_graph_convolution_pyg.py`) |
| Potential wrapper | `matgl.apps._pes_dgl.Potential` | `matgl.apps._pes_pyg.Potential` |
| Structure converter | `matgl.ext._pymatgen_dgl.Structure2Graph` | `matgl.ext._pymatgen_pyg.Structure2Graph` |

Weight transfer is exact: `state.pt` from the DGL checkpoint is loaded without
modification into the PyG model, since all `nn.Module` parameter names are preserved.

---

## Parity Test

An automated pytest verifies that the PyG model reproduces DGL reference values
to within `atol = 1e-5` on four structures per functional:

```
tests/models/test_chgnet_pyg.py::test_matpes_model_parity_pyg
```

**Status: 8/8 passed** (PBE × {MoS, Fe, MoS_perturbed, Fe_perturbed} +
r2SCAN × same four structures).

To run:
```bash
pytest tests/models/test_chgnet_pyg.py::test_matpes_model_parity_pyg -v
```

---

## Extended Parity Evaluation

Both models (PBE and r2SCAN) were evaluated on 10 structures spanning a range of
chemistries and coordination environments.  Predictions were generated independently
with `MATGL_BACKEND=DGL` and `MATGL_BACKEND=PYG` using
`scripts/predict_chgnet_parity.py` and compared element-wise.

### Structures tested

| ID | Formula | Description |
|---|---|---|
| `mos` | MoS | 2-atom cubic (a = 4.00 Å) |
| `fe` | Fe₂ | 2-atom BCC (a = 2.87 Å) |
| `mos_perturbed` | MoS | Same, small fractional displacement |
| `fe_perturbed` | Fe₂ | Same, small fractional displacement |
| `mo` | Mo | 1-atom BCC (a = 3.15 Å) |
| `al` | Al | 1-atom FCC (a = 4.05 Å) |
| `nacl` | NaCl | 2-atom rocksalt (a = 5.64 Å) |
| `batio3` | BaTiO₃ | 5-atom cubic perovskite (a = 4.00 Å) |
| `li2o` | Li₂O | 3-atom antifluorite (a = 4.62 Å) |
| `mgo` | MgO | 2-atom rocksalt (a = 4.21 Å) |

### Results — PBE

| Structure | \|ΔE/atom\| (eV) | max\|ΔF\| (eV/Å) | max\|ΔS\| (GPa) | max\|Δmag\| (μ_B) |
|---|---|---|---|---|
| mos | 0 | 0 | 0 | 0 |
| fe | 0 | 7.1×10⁻⁸ | 2.9×10⁻⁶ | 0 |
| mos\_perturbed | 0 | 0 | 0 | 0 |
| fe\_perturbed | 0 | 6.0×10⁻⁸ | 1.9×10⁻⁶ | 0 |
| mo | 0 | 0 | 0 | 0 |
| al | 0 | 0 | 0 | 0 |
| nacl | 0 | 0 | 0 | 0 |
| batio3 | 0 | 1.5×10⁻⁷ | 4.1×10⁻⁶ | 0 |
| li2o | 0 | 0 | 0 | 0 |
| mgo | 0 | 0 | 0 | 0 |
| **worst case** | **0** | **1.5×10⁻⁷** | **4.1×10⁻⁶** | **0** |

### Results — r2SCAN

| Structure | \|ΔE/atom\| (eV) | max\|ΔF\| (eV/Å) | max\|ΔS\| (GPa) | max\|Δmag\| (μ_B) |
|---|---|---|---|---|
| mos | 0 | 0 | 0 | 0 |
| fe | 0 | 7.5×10⁻⁸ | 1.1×10⁻⁶ | 0 |
| mos\_perturbed | 0 | 0 | 0 | 0 |
| fe\_perturbed | 0 | 1.5×10⁻⁷ | 1.1×10⁻⁶ | 0 |
| mo | 0 | 0 | 0 | 0 |
| al | 0 | 0 | 0 | 0 |
| nacl | 0 | 0 | 0 | 0 |
| batio3 | 0 | 2.7×10⁻⁷ | 4.2×10⁻⁶ | 0 |
| li2o | 0 | 0 | 0 | 0 |
| mgo | 0 | 0 | 0 | 0 |
| **worst case** | **0** | **2.7×10⁻⁷** | **4.2×10⁻⁶** | **0** |

### Interpretation

- **Energy and magnetic moments** agree to exact float32 equality (difference = 0) on all
  10 structures for both functionals.
- **Forces and stresses** show non-zero differences only on structures with genuinely
  non-zero forces/stresses (cubic high-symmetry structures are zero by symmetry and
  match exactly).  The maximum discrepancy is **2.7×10⁻⁷ eV/Å** for forces and
  **4.2×10⁻⁶ GPa** for stresses — both well below float32 machine epsilon
  (ε ≈ 1.2×10⁻⁷) accumulated over the autograd graph.  These differences arise from
  different summation order in the DGL vs. PyG line-graph message-passing kernels and
  are not physically meaningful.
- All differences are **more than two orders of magnitude below** the `atol = 1e-5`
  threshold used by the automated parity test.

---

## Reproducibility

The automated pytest is the canonical regression guard:

```bash
pytest tests/models/test_chgnet_pyg.py::test_matpes_model_parity_pyg -v
```

The extended 10-structure evaluation above was generated with a local dev script
(not checked in). The parity numbers reported here are the reference values embedded
in the automated test.

### CODE REVIEW

**Comment by shyuep** [comment]
**🤖 Automated PR review (generated by Claude on behalf of @shyuep)**

Well-documented PyG port of CHGNet with direct weight transfer from the DGL checkpoints (no retraining). Parity tests (`test_compute_parity.py`, `test_graph_conv_chgnet_pyg.py`, `test_chgnet_pyg.py`) and the `test_pyg` CI matrix all pass — good coverage for a backend port. A few items before merge:

- ⚠️ **Misleading comment in `CHGNetGraphConv.forward`** (`_graph_convolution_pyg.py`): the inline comment reads `# Node update — scatter onto src (central atoms)`, but `node_update_` scatters onto `dst` (`scatter_add(messages, dst, ...)`) — which its own docstring correctly states matches DGL `fn.sum`. The implementation is right; the `forward` comment is stale and contradicts it. In a parity-critical port this is dangerous — please correct the comment.
- ⚠️ **Perf regression in `_ase_pyg.py` `calculate`**: the refactor dropped the `host = [t.detach().cpu() ...]` batching and now calls `calc_result[0].detach().cpu().numpy().item()` twice (for `energy` and `free_energy`). The deleted code's comment explicitly warned that repeated `.detach().cpu()` on the same tensor triggers an extra host sync per MD step. Compute it once and reuse.
- 🔸 **Docstring indentation**: in `CHGNet.__init__` (`_chgnet_pyg.py`) and `Relaxer.__init__` (`_ase_pyg.py`) the `Args:` entries are not indented under the `Args:` header, which breaks Google-style docstring rendering. Re-indent for consistency with the rest of the codebase.
- 🔸 **CI**: `pre-commit.ci` (`ruff-check`) is failing — it flags RUF003 ambiguous-dash characters, and the new comments/docstrings here use `—`, `→`, `–`. Please run `ruff check` and resolve before merge (part of the failure is also pre-existing repo-wide lint debt).

Message-passing semantics and architecture otherwise look correct and faithful to the DGL implementation.

**Comment by bowen-bd** [comment]
@shyuep thanks for the review! The concerns should be resolved now

**Comment by shyuep** [comment]
Thanks! Yes, if you would like, pls transfer the models to materialyze org. One minor thing I would ask is to rename the PyG in your model names to PES. We are standardizing all models to `<architecture>-<purpose>-<dataset+version>` format.

**Comment by bowen-bd** [comment]
@shyuep I think it will be great to upload these checkpoints to materialyze HF. However I'm not a member of the org so can not do this. Can you please download and rename these models and upload to materialyze HF?

**Comment by shyuep** [comment]
Thanks. I just simply uploaded and override them in the existing models.
